File: Sources/GatewayCore/Voice.swift
import Foundation

/// Engine settings for the voice. As of the v4 fork this is a fixed,
/// fine-tuned voice bundled with the app -- not a runtime clone built from a
/// user-supplied reference recording the way Qwen3's ICL conditioning was.
/// `referenceWav`/`referenceText` are kept only so an old v3 `profile.json`
/// still decodes without throwing; nothing in v4 writes or reads them for
/// rendering. `modelVersion` is the field that actually matters now: bump it
/// whenever the bundled `.onnx` is replaced, so `renderKey` correctly
/// invalidates old takes the same way a changed reference clip used to.
public struct VoiceProfile: Codable, Sendable, Equatable {
    public var engine: String
    /// Bumped when the bundled model file changes. This, not a reference
    /// clip, is now the thing that actually changes rendered audio.
    public var modelVersion: String
    @available(*, deprecated, message: "unused since the v4 fork; kept only so old profile.json files still decode")
    public var referenceWav: String
    @available(*, deprecated, message: "unused since the v4 fork; kept only so old profile.json files still decode")
    public var referenceText: String
    /// Reference QC target for spectral tilt, in dB. Meaningless for a
    /// bundled fixed voice -- kept for the same decode-compatibility reason.
    @available(*, deprecated, message: "unused since the v4 fork; kept only so old profile.json files still decode")
    public var targetAlphaDB: Double

    public init(engine: String = Engine.name, modelVersion: String = "1",
                referenceWav: String = "reference.wav", referenceText: String = "",
                targetAlphaDB: Double = -14) {
        self.engine = engine; self.modelVersion = modelVersion
        self.referenceWav = referenceWav
        self.referenceText = referenceText; self.targetAlphaDB = targetAlphaDB
    }

    /// Hand-editable JSON: a missing key falls back instead of throwing, the
    /// same rule levels.json follows.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        engine = try c.decodeIfPresent(String.self, forKey: .engine) ?? Engine.name
        modelVersion = try c.decodeIfPresent(String.self, forKey: .modelVersion) ?? "1"
        referenceWav = try c.decodeIfPresent(String.self, forKey: .referenceWav) ?? "reference.wav"
        referenceText = try c.decodeIfPresent(String.self, forKey: .referenceText) ?? ""
        targetAlphaDB = try c.decodeIfPresent(Double.self, forKey: .targetAlphaDB) ?? -14
    }

    /// Only the fields that change rendered audio: the engine, and which
    /// version of its bundled model. Cache keys are backend-aware (v1
    /// lesson): hash this, not the whole profile.
    public var renderKey: String { "\(engine)|\(modelVersion)" }
}

public enum VoiceProfileIO {
    public static func load(from url: URL) -> VoiceProfile {
        guard let data = try? Data(contentsOf: url),
              let p = try? JSONDecoder().decode(VoiceProfile.self, from: data)
        else { return VoiceProfile() }
        return p
    }

    public static func save(_ profile: VoiceProfile, to url: URL) throws {
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(),
                                                withIntermediateDirectories: true)
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        try enc.encode(profile).write(to: url, options: .atomic)
    }
}

File: cross-platform/src/core/voice.ts
/**
 * The voice: its engine settings, where its folder lives, and which one a
 * session actually renders with.
 *
 * `VoiceProfile` is a *fixed, bundled* voice as of the v4 fork — not a runtime
 * clone built from a user-supplied reference. `referenceWav`/`referenceText`
 * are kept only so an old v3 `profile.json` still decodes; nothing reads them
 * for rendering. `modelVersion` is the field that matters now: bump it when
 * the bundled model is replaced, so `renderKey` invalidates old takes the same
 * way a changed reference clip used to.
 */

/** The engine's own name. Not probed here: `Engine.missingResourceParts`
 *  inspects an app bundle, which is the one seam this port deliberately does
 *  not cross — so clonability is *supplied* to the resolver below rather than
 *  discovered. */
import { join } from "path";

export const engineName = "piper-snepssen";

export interface VoiceProfile {
  engine: string;
  /** Bumped when the bundled model file changes. This, not a reference clip,
   *  is now the thing that actually changes rendered audio. */
  modelVersion: string;
  referenceWav: string;
  referenceText: string;
  targetAlphaDB: number;
}

export const defaultProfile = (): VoiceProfile => ({
  engine: engineName, modelVersion: "1", referenceWav: "reference.wav",
  referenceText: "", targetAlphaDB: -14,
});

/** Hand-editable JSON: a missing key falls back instead of throwing, the same
 *  rule levels.json follows. A malformed file is the default profile, not an
 *  error — the Swift `try?` swallows both cases identically. */
export function decodeProfile(raw: unknown): VoiceProfile {
  const d = defaultProfile();
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return d;
  const o = raw as Record<string, unknown>;
  return {
    engine: typeof o.engine === "string" ? o.engine : d.engine,
    modelVersion: typeof o.modelVersion === "string" ? o.modelVersion : d.modelVersion,
    referenceWav: typeof o.referenceWav === "string" ? o.referenceWav : d.referenceWav,
    referenceText: typeof o.referenceText === "string" ? o.referenceText : d.referenceText,
    targetAlphaDB: typeof o.targetAlphaDB === "number" ? o.targetAlphaDB : d.targetAlphaDB,
  };
}

/** Only the fields that change rendered audio: the engine, and which version
 *  of its bundled model. Cache keys are backend-aware — hash this, not the
 *  whole profile. */
export const renderKey = (p: VoiceProfile): string => `${p.engine}|${p.modelVersion}`;

// ------------------------------------------------------------ the folder

/** Folders starting `_` are working areas, not voices — `_audition` is one. */
export function isValidName(name: string): boolean {
  if (name === "") return false;
  if (name.startsWith("_")) return false;
  // Swift's `isLetter`/`isNumber` are Unicode properties, not ASCII: a voice
  // called `café` or `声` is a valid folder name on both sides.
  return [...name].every(c => /\p{L}/u.test(c) || /\p{N}/u.test(c) || c === "-" || c === "_");
}

/** One short line that exercises what a listener will actually hear: a
 *  sentence, a silence, and whether the voice comes back the same. */
export const previewText =
  "Rest here for a moment, and let your breathing settle. "
  + "When you are ready, we will go on.";

// ------------------------------------------------------------ the folder

export const voicesRoot = (root: string): string => join(root, "voices");
export const voiceDir = (root: string, name: string): string => join(voicesRoot(root), name);
export const previewURL = (root: string, name: string): string =>
  join(voiceDir(root, name), "preview.wav");
/** The preview is stale when the thing it previews has changed. */
export const previewStampURL = (root: string, name: string): string =>
  join(voiceDir(root, name), "preview.engine");
export const profileURL = (root: string, name: string): string =>
  join(voiceDir(root, name), "profile.json");
